Implementation of a multiservice for backend services

flongo opened the pull request:
Mode
- creation of the multiservice method
- conversion of the "validate" method to the new pattern

Code
- creation of a cache mechanism that prevents the execution of a service in case it has been executed on the same source code

Validation visitor
- the code to perform the validation of the graph with the extraction of errors and warnings has been moved to this dedicated class

Partial fix of issue #50

ymeine commented on app/node_modules/modes/node_modules/mode.js:
```diff
@@ -103,11 +100,67 @@ var Mode = oop.Class({
 
 
+		/**
+		 * Unique service that callows to call several of them with a unique node traversal
```
typo: `callows`

ymeine commented on app/node_modules/modes/node_modules/mode.js:
```diff
@@ -103,11 +100,67 @@ var Mode = oop.Class({
+		 */
+		multiService: function(code, options) {
+			var graph = code.graph, response = {};
+			if (!graph) {
```
Just for consistency: `if (graph == null)` instead

ymeine commented on app/node_modules/modes/node_modules/mode.js:
```diff
@@ -103,11 +100,67 @@ var Mode = oop.Class({
+				 return response;
+			}
+
+			for (var vis in options) {
```
Please add a comment explaining you're fetching every cached results and removing the corresponding visitors in order not to run them afterwards.
Also, rename `vis` to `visitor`

ymeine commented on app/node_modules/modes/node_modules/mode.js:
```diff
@@ -103,11 +100,67 @@ var Mode = oop.Class({
+
+			for (var vis in options) {
+				if (options.hasOwnProperty(vis) && code.getCache(vis)) {
+					response[vis] = code.getCache(vis);
```
Avoid duplicate call of the same method, assign the first result to a variable, and compare the latter with `null` (explicitly showing that the result is not a boolean but the actual cached result if available)

flongo replied:
I don't think it is a big deal...

ymeine commented on app/node_modules/modes/node_modules/mode.js:
```diff
@@ -103,11 +100,67 @@ var Mode = oop.Class({
+		 * @param {Object} visitorsToCall keys represent the visitors to call, values are the extra parameters to pass to the methods of the visitors
+		 * @param {String} method name of the method to call. Visitors implement 4 methods: "start", "emptyOutput", "processNode", "end". "processNode" method receives the node that is currently being traversed as an argument
+		 * @param {Object} args First argument of the call
+		 * @param {Function} callback Function to be called at the end of the processing. It receives as first argument the value returned by the method
```
The first argument is the visitor name, the result is passed as the second one.

ymeine commented on app/node_modules/modes/node_modules/mode.js:
```diff
@@ -103,11 +100,67 @@ var Mode = oop.Class({
+		callVisitors: function(visitorsToCall, method, args, callback) {
+			var result;
+			for (var vis in visitorsToCall) {
+				if (visitorsToCall.hasOwnProperty(vis) && visitors[vis]) {
```
Idem: `visitors[vis] != null`

ymeine commented on app/node_modules/modes/node_modules/mode.js:
```diff
@@ -103,11 +100,67 @@ var Mode = oop.Class({
+				graph.traverse(function(node) {
+					that.callVisitors.apply(that, [options, "processNode", node])
+				});
+				this.callVisitors(options, "end", null, function(vName, result) {
```
`vName` :arrow_right: `visitorName`

ymeine commented on app/node_modules/modes/node_modules/mode.js:
```diff
@@ -103,11 +100,67 @@ var Mode = oop.Class({
+				this.callVisitors(options, "start", graph);
+				var that = this;
+				graph.traverse(function(node) {
+					that.callVisitors.apply(that, [options, "processNode", node])
```
I guess using `apply` here is not necessary, since `callVisitors` should already receive as scope its caller, that is `that` ( :stuck_out_tongue_winking_eye: )

flongo replied:
I don't understand the remark :)

ymeine commented on app/node_modules/modes/node_modules/validation/visitor.js:
```diff
@@ -0,0 +1,62 @@
+		 */
+		processNode: function(node) {
+			var err = node.get('errors');
+			if (err) {
```
Full names: `err` :arrow_right: `errors`
and explicit comparisons: `errors != null`

ymeine commented on app/node_modules/modes/node_modules/validation/visitor.js:
```diff
@@ -0,0 +1,62 @@
+					messages: err
+				});
+			}
+			var warn = node.get('warnings');
```
Idem: 
- `warn` :arrow_right: `warnings`
- `warnings != null`

ymeine commented on app/node_modules/modes/node_modules/code.js:
```diff
@@ -128,21 +129,26 @@ var Code = oop.class({
+
+				if (this.source == "") {
+					this._graph = null;
+					return {
```
This statement is repeated below, just move the test inside the `try` block (you can even use a ternary expression here to assign `this._graph`)

ymeine commented on app/node_modules/modes/node_modules/code.js:
```diff
@@ -83,6 +83,7 @@ var Code = oop.class({
 
 		this._source = '';
 		this._graph = undefined;
+		this._cache = {};
```
Considering only this class, without external context, it's hard to figure out what `cache` means. It could be anything. Maybe rename it to a more explicit variable: like `_servicesCache`

I know it can be an attempt of making it generic, but in this case I would expect a totally external implementation. And also mixing data of different purposes inside the same cache is not that easy with a one level deep object.

flongo committed:
Implementation of a multiservice

Mode
- creation of the multiservice method
- conversion of the "validate" method to the new pattern

Code
- creation of a cache mechanism that prevents the execution of a service in case it has been executed on the same source code

Validation visitor
- code to perform the validation of the graph with the extraction of errors and warnings has been moved to this dedicated class

Partial fix of issue #50

flongo commented:
Integrated in ba57366773be985771fea51e83de2db619d38fd9
